From b1d77fac5c0dcd6248058f38fdeb1e0052bc7790 Mon Sep 17 00:00:00 2001
From: Bin Jiang <bin.jiang@windriver.com>
Date: Thu, 20 Nov 2014 17:13:22 +0800
Subject: [PATCH] netdev: octeon-bgx-port: bgx_port_check_state work can't be
 used for XAUI.

bgx_port_check_state work can't used for XAUI, so don't queue
bgx_port_check_state work for XAUI.

bgx_port_check_state work should not be queued repeatedly, so add a
condition for it.

Signed-off-by: Bin Jiang <bin.jiang@windriver.com>
---
 drivers/net/ethernet/octeon/octeon-bgx-port.c |   11 ++++++++---
 1 files changed, 8 insertions(+), 3 deletions(-)

diff --git a/drivers/net/ethernet/octeon/octeon-bgx-port.c b/drivers/net/ethernet/octeon/octeon-bgx-port.c
index 6701ca2..e6b4a61 100644
--- a/drivers/net/ethernet/octeon/octeon-bgx-port.c
+++ b/drivers/net/ethernet/octeon/octeon-bgx-port.c
@@ -365,6 +365,9 @@ int bgx_port_enable(struct net_device *netdev)
 		cvmx_helper_link_autoconf(priv->ipd_port);
 		netif_carrier_on(netdev);
 
+		if (cvmx_helper_interface_get_mode(priv->xiface) == CVMX_HELPER_INTERFACE_MODE_XAUI)
+			return 0;
+
 		mutex_lock(&check_state_wq_mutex);
 		if (!check_state_wq) {
 			check_state_wq =
@@ -376,9 +379,11 @@ int bgx_port_enable(struct net_device *netdev)
 			return -ENOMEM;
 
 		mutex_lock(&priv->lock);
-		INIT_DELAYED_WORK(&priv->dwork, bgx_port_check_state);
-		queue_delayed_work(check_state_wq, &priv->dwork, 0);
-		priv->work_queued = true;
+		if(priv->dwork.work.func != bgx_port_check_state) {
+			INIT_DELAYED_WORK(&priv->dwork, bgx_port_check_state);
+			queue_delayed_work(check_state_wq, &priv->dwork, 0);
+			priv->work_queued = true;
+		}
 		mutex_unlock(&priv->lock);
 
 		pr_info("%s: Link is not ready\n", netdev->name);
-- 
1.7.5.4

